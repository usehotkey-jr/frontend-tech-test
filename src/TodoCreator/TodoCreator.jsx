// @flow

import React from "react";
import {connect} from "react-redux";
import type {Todo} from "../newTodo/newTodo.duck";
import {newTodoActions} from "../newTodo/newTodo.duck";
import {selectNewTodo} from "../newTodo/newTodo.selector";
import {Button, Input} from "./TodoCreator.styled";

type TodoCreatorStateProps = {
    newTodo: Todo;
};

type TodoCreatorProps = typeof newTodoActions & TodoCreatorStateProps;

export class TodoCreatorComponent extends React.PureComponent<TodoCreatorProps> {
    makeOnChange = (key: string) => (event: SyntheticEvent<HTMLInputElement>) => {
        if (this.props.change) {
            this.props.change([key, event.currentTarget.value]);
        }
    };

    render () {
        const {newTodo} = this.props;

        return (
            <div>
                <Input onChange={this.makeOnChange("title")} value={newTodo.title}/>
                <Input onChange={this.makeOnChange("description")} value={newTodo.description}/>
                <Button onClick={this.props.add}>Add TODO</Button>
            </div>
        );
    }
}

export const TodoCreator = connect(
    selectNewTodo,
    newTodoActions
)(TodoCreatorComponent);
