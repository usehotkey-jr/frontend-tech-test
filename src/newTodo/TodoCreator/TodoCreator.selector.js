// @flow

import {selectNewTodo} from "../newTodo.selector";
import {createSelector} from "reselect";

export const selectTodoCreatorProps = createSelector(
    selectNewTodo,
    (newTodo) => ({newTodo})
);
