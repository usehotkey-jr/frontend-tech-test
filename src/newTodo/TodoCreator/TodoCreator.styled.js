// @flow

import styled from "styled-components";
import {palette} from "../../theme/palette";
import {size} from "../../theme/size";

const {small, medium, large} = size;
const {fg, stroke, boxShadow} = palette;

export const ContainerTC = styled.div`
    padding: ${medium.spacing};
`;

export const BlockContainerTC = styled.div`
    margin-bottom: ${medium.spacing};
`;

export const InputTC = styled.input`
    width: 100%;
    padding: ${small.spacing} ${large.spacing};
    
    color: ${fg.primary};
    
    border-radius: ${small.radius};  
    border: 1px solid ${stroke.primary};  
    box-shadow: ${boxShadow.primaryInner};
    
    &:focus {
    border: 1px solid ${stroke.control}; 
    box-shadow: ${boxShadow.primaryFocused};  
    }
`;

export const TextAreaTC = InputTC.withComponent("textarea").extend`
    resize: vertical;
`;
