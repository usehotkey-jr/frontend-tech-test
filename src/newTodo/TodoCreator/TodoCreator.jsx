// @flow

import React from "react";
import {connect} from "react-redux";
import type {Todo} from "../newTodo.duck";
import {newTodoActions} from "../newTodo.duck";
import {BlockContainerTC, ContainerTC} from "./TodoCreator.styled";
import {createNewTodo} from "../newTodo.thunk";
import {selectTodoCreatorProps} from "./TodoCreator.selector";
import {Button} from "../../_components/Button.styled";
import {TextArea} from "../../_components/TextArea.styled";
import {Input} from "../../_components/Input.styled";

type TodoCreatorStateProps = {
    newTodo: Todo;
};

type TodoCreatorDispatchProps = {
    change: typeof newTodoActions.change;
    createNewTodo: typeof createNewTodo;
};

type TodoCreatorProps = TodoCreatorStateProps & TodoCreatorDispatchProps;

export class TodoCreatorComponent extends React.PureComponent<TodoCreatorProps> {
    makeOnChange = (key: string) =>
        (event: SyntheticEvent<HTMLInputElement>) =>
            this.props.change([key, event.currentTarget.value]);

    render () {
        const {newTodo} = this.props;
        const isAddDisabled = !newTodo.title || !newTodo.description;

        return (
            <ContainerTC>
                <BlockContainerTC>
                    <Input
                        onChange={this.makeOnChange("title")}
                        value={newTodo.title}
                        placeholder="Title"
                    />
                </BlockContainerTC>
                <BlockContainerTC>
                    <TextArea
                        onChange={this.makeOnChange("description")}
                        value={newTodo.description}
                        placeholder="Description"
                    />
                </BlockContainerTC>
                <Button
                    onClick={this.props.createNewTodo}
                    disabled={isAddDisabled}
                >
                    Add task
                </Button>
            </ContainerTC>
        );
    }
}

export const TodoCreator = connect(
    selectTodoCreatorProps,
    {
        change: newTodoActions.change,
        createNewTodo,
    }
)(TodoCreatorComponent);
