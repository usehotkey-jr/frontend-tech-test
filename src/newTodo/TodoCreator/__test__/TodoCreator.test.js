import {createTestSnapshot} from "../../../_helpers/jest/createTestSnapshot.ignore";
import {TodoCreatorComponent} from "../TodoCreator";
import {todoMock} from "../../../todoList/__mocks__/todos.mock";

describe("TodoCreator", () => {
    test("should correctly render", () => {
        const props = {
            newTodo: todoMock,
            change: jest.fn(),
            createNewTodo: jest.fn(),
        };
        const wrapper = createTestSnapshot.all(TodoCreatorComponent, props);

        // Title
        const titleInput = wrapper.find("input");
        expect(titleInput.props().value).toBe(todoMock.title);

        titleInput.simulate("change");
        expect(props.change.mock.calls).toHaveLength(1);

        // Description
        const descriptionTextarea = wrapper.find("textarea");
        expect(descriptionTextarea.props().value).toBe(todoMock.description);

        descriptionTextarea.simulate("change");
        expect(props.change.mock.calls).toHaveLength(2);

        // Add button
        wrapper.find("button").simulate("click");
        expect(props.createNewTodo.mock.calls).toHaveLength(1);
    });
});
