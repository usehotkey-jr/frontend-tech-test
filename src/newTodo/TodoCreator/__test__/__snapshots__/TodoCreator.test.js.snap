// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoCreator should correctly render 1`] = `
<styled.div>
  <styled.div>
    <styled.input
      onChange={[Function]}
      placeholder="Title"
      value="Hi"
    />
  </styled.div>
  <styled.div>
    <styled.textarea
      onChange={[Function]}
      placeholder="Description"
      value="Guys"
    />
  </styled.div>
  <styled.button
    disabled={false}
    onClick={[Function]}
  >
    Add task
  </styled.button>
</styled.div>
`;

exports[`TodoCreator should correctly render 2`] = `
<div
  class="sc-bdVaJa btOCPD"
>
  <div
    class="sc-bwzfXH igYdVz"
  >
    <input
      class="sc-bxivhb bStphO"
      placeholder="Title"
      value="Hi"
    />
  </div>
  <div
    class="sc-bwzfXH igYdVz"
  >
    <textarea
      class="sc-EHOje kyhFuH"
      placeholder="Description"
    >
      Guys
    </textarea>
  </div>
  <button
    class="sc-htpNat jOuzPi"
  >
    Add task
  </button>
</div>
`;

exports[`TodoCreator should correctly render 3`] = `
<TodoCreatorComponent
  change={[Function]}
  createNewTodo={[Function]}
  newTodo={
    Object {
      "description": "Guys",
      "id": 0,
      "title": "Hi",
    }
  }
>
  <styled.div>
    <div
      className="sc-bdVaJa btOCPD"
    >
      <styled.div>
        <div
          className="sc-bwzfXH igYdVz"
        >
          <styled.input
            onChange={[Function]}
            placeholder="Title"
            value="Hi"
          >
            <input
              className="sc-bxivhb bStphO"
              onChange={[Function]}
              placeholder="Title"
              value="Hi"
            />
          </styled.input>
        </div>
      </styled.div>
      <styled.div>
        <div
          className="sc-bwzfXH igYdVz"
        >
          <styled.textarea
            onChange={[Function]}
            placeholder="Description"
            value="Guys"
          >
            <textarea
              className="sc-EHOje kyhFuH"
              onChange={[Function]}
              placeholder="Description"
              value="Guys"
            />
          </styled.textarea>
        </div>
      </styled.div>
      <styled.button
        disabled={false}
        onClick={[Function]}
      >
        <button
          className="sc-htpNat jOuzPi"
          disabled={false}
          onClick={[Function]}
        >
          Add task
        </button>
      </styled.button>
    </div>
  </styled.div>
</TodoCreatorComponent>
`;
